_source/_posts/2021-10-27-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,565 @@
+  "idp": "0oa28lrldhsBglcC35d7",
+  "nonce": "2Rsv7E84vo_K14J6qm9IgKjNsCUxb4Hc16qgzx9CKlk",
+  "auth_time": 1634649927,
+  "at_hash": "veXQkvqTjCl70OkZnKZVyA",
+  "groups": [
+    "k8s-admins"
+  ]
+}
+
+```
+
+Now let us update kubectl configuration to add an OIDC user:
+
+```bash
+kubectl config set-credentials oidc-user \
+--exec-api-version=client.authentication.k8s.io/v1beta1 \
+--exec-command=kubectl \
+--exec-arg=oidc-login \
+--exec-arg=get-token \
+--exec-arg=--oidc-issuer-url=<k8s_oidc_issuer_url> \
+--exec-arg=--oidc-client-id=<k8s_oidc_client_id> \
+--exec-arg=--oidc-extra-scope="email offline_access profile openid"
+```
+
+Now we can use this user to authenticate to the cluster via kubectl.
+
+```bash
+kubectl get pods --user=oidc-user -n default
+```
+
+This should open a browser window to authenticate you. You can use your Okta account credentials to log in.
+
+You can play around with different users in different groups to test out the RBAC configuration we did. Users who are not in the `k8s-admins` group in Okta should not be able to access any resource. Users who are only in the `restricted-users` group should only be able to see pods and services.
+
+You can use the below command to set this user as default for your current kubectl context.
+
+```
+kubectl config set-context --current --user=oidc-user
+```
+
+# Learn more about using OIDC with Kubernetes
+
+Using OIDC is a great way to secure your Kubernetes clusters, especially in bigger teams. It is more secure than any of the default Kubernetes authentication mechanisms. On top of that, it will let you manage users and their roles and permissions in your cluster and even add multi-factor authentication for your cluster.
+
+While this exercise shows how to do it with Okta, the process would be very similar for other OIDC providers as well. Check out the [Okta Terraform provider docs](https://registry.terraform.io/providers/okta/okta/latest/docs) to see what else can be automated via Terraform. You can also automate the Kubernetes parts via terraform using the [Kubernetes Terraform provider](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs).
+
+If you want to learn more about Kubernetes, OIDC, or using OIDC with Kubernetes, feel free to follow the links below.
+
+- [OAuth 2.0 and OpenID Connect Overview](https://developer.okta.com/docs/concepts/oauth-openid/)
+- [Secure Access to AWS EKS Clusters for Admins](https://developer.okta.com/blog/2021/10/08/secure-access-to-aws-eks)
+- [Build a Microservice Architecture with Spring Boot and Kubernetes](https://developer.okta.com/blog/2019/04/01/spring-boot-microservices-with-kubernetes)
+- [Kubernetes to the Cloud with Spring Boot and JHipster](https://developer.okta.com/blog/2021/06/01/kubernetes-spring-boot-jhipster)
+- [Using Okta Advanced Server Access & Terraform to Automate Identity & Infrastructure as Code](https://developer.okta.com/blog/2020/04/24/okta-terraform-automate-identity-and-infrastructure)
```

Comment:
We tend to use relative links to blog posts, so these URLs should start with `/blog`. You can run `npm run clean-post` to do this with a script.

Reply:
That's strange, I see there is a pre-commit hook that's supposed to catch this right? may be we can just run the command as part of pre-commit?

Reply:
This is starnge
```
❯ npm run clean-post

> @oktadev/okta-blog@1.0.0 clean-post /home/deepu/Workspace/okta/okta-blog
> node scripts/dev-chars

Reading latest post 2021-11-02-k8s-to-the-cloud-aws.md
Found 0 curly quotes, 0 absolute blog URLs, and 0 'GitHub' case issues.
Replace characters? [Y/n] y

Replaced 0 characters
Replaced 0 instances of absolute blog URLs
Replaced 0 instances of 'Github' to 'GitHub'
```

Reply:
ah damn, didn't realize this only reads latest post. I need to update the date :facepalm: